File: Templates/Day_3/Reduction/SampleSolution/C/pi_square_omp_reduction.c
#include <stdio.h>
#ifdef _OPENMP
#include <omp.h>
#endif

int main()
{

  const int finval = 10000;

  double pi_square = 0.0;

#ifdef _OPENMP
  double start_time = omp_get_wtime();
#endif

#pragma omp parallel for default(none) reduction(+:pi_square)  
  for ( int i=1 ; i <= finval; i++)
    {  
      double factor = (double) i;
      pi_square += 1.0/( factor * factor );
    }
#ifdef _OPENMP
  double end_time = omp_get_wtime() - start_time;
#else
  double end_time = 0.0;
#endif

  printf ( "Pi^2 = %.10f\n", pi_square*6.0);
  printf ( "Time taken: %f s\n", end_time);

  return 0;

}

File: Templates/Day_3/Reduction/SampleSolution/Fortran/pi_square_omp_reduction.f90
Program pi

  !$ use omp_lib
  
  implicit none

  integer, parameter :: finval = 10000
  
  double precision :: pi_square = 0.0d0

  ! timing
  !$ double precision start_time, end_time
  
  
  double precision :: factor
  integer:: i

  !$ start_time = omp_get_wtime()

  !$omp parallel do reduction(+:pi_square) private(factor)
  do i= 1, finval
     factor = i
     pi_square = pi_square + 1.0d0/(factor * factor)
  enddo
  !$omp end parallel do

  !$ end_time = omp_get_wtime() - start_time
  
  print *, 'Pi**2 =', 6.0d0 * pi_square
  !$ print *, 'Time taken: ', end_time
  
end Program pi
